smarty/environment.py

```diff
@@ -465,70 +493,132 @@ def _asSMIRKS(self, initialAtom = None, neighbors = None, smarts = False):
+        or None if no such atom exists
         """
-        if atomIndex == None:
+        if descriptor == None:
             return random.choice(self._graph.nodes())
-        atoms = self.getAtoms()
-        indices = [a.index for a in atoms]
-        if atomIndex in indices:
-            return atoms[indices.index(atomIndex)]
-        # TODO: index is not specified raise error instead?
-        else:
+
+        try: descriptor = int(descriptor)
+        except: descriptor = descriptor
+
+        if type(descriptor) is int:
+            for atom in self.getAtoms():
+                if atom.index == descriptor:
+                    return atom
             return None
 
-    def selectBond(self, atomIndex1 = None, atomIndex2 = None):
-        """Select a bond with uniform probability.
-        Bonds are found by the two atoms that define them
+        atoms = self.getComponentList('atom',descriptor)
+        if len(atoms) == 0:
+            return None
+
+        return random.choice(atoms)
+
+    def getComponentList(self, component_type, descriptor = None):
+        """
+        Returns a list of atoms or bonds matching the descriptor
 
         Parameters
-        ----------
-        atomIndex1 and atomIndex2: int, optional
-            indices for atoms framing the bond
+        -----------
+        component_type: string: 'atom' or 'bond'
+        descriptor: string, optional
+            'all', 'Indexed', 'Unindexed', 'Alpha', 'Beta'
+
+        Returns
+        -------
+        """
+        if descriptor is not None:
+            d = descriptor.lower()
+        else:
+            d = None
+
+        if not component_type.lower() in ['atom', 'bond']:
+            raise Exception("Error: 'getComponentList()' component_type must be 'atom' or 'bond'")
+
+        if component_type.lower() == 'atom':
+            if d == 'indexed':
+                return self.getIndexedAtoms()
+            elif d == 'unindexed':
+                return self.getUnindexedAtoms()
+            elif d == 'alpha':
+                return self.getAlphaAtoms()
+            elif d == 'beta':
+                return self.getBetaAtoms()
+            else:
+                return self.getAtoms()
+
+        elif component_type.lower() == 'bond':
+            if d == 'indexed':
+                return self.getIndexedBonds()
+            elif d == 'unindexed':
+                return self.getUnindexedBonds()
+            elif d == 'alpha':
+                return self.getAlphaBonds()
+            elif d == 'beta':
+                return self.getBetaBonds()
+
+            return self.getBonds()
+
+        return None
+
+    def selectBond(self, descriptor = None):
+        """Select a random bond fitting the descriptor.
+
+        Paramters
+        ---------
+        descriptor: optional, None
+            None - returns any bond with equal probability
+            int - will return an bond with that index
+            'Indexed' - returns a random indexed bond
```

Comment:
Do we have indexed bonds? I thought we only had indexed atoms? (Or does "indexed bond" mean a bond between indexed atoms?)